### PR TITLE
add sifive_uart

### PR DESCRIPTION
**Pull request by RubBra**

### CODE REVIEW

File: src/vcml/models/serial/sifive_uart.cpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    const u32 mask = (TXCTRL_TXEN | TXCTRL_NSTOP | TXCTRL_TXCNT());
+    txctrl = val & mask;
+    if (is_tx_enabled())
+        send_tx();
+    update_tx();
+}
+
+void sifive_uart::write_rxctrl(u32 val) {
+    rxctrl = val & (RXCTRL_RXEN | RXCTRL_NSTOP | RXCTRL_RXCNT());
+    update_tx();
```
Comment (janweinstock) [inline src/vcml/models/serial/sifive_uart.cpp:156]: why update tx during receiver control write?

File: src/vcml/models/serial/sifive_uart.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    return val;
+}
+
+void sifive_uart::write_txctrl(u32 val) {
+    const u32 mask = (TXCTRL_TXEN | TXCTRL_NSTOP | TXCTRL_TXCNT());
+    txctrl = val & mask;
+    if (is_tx_enabled())
+        send_tx();
+    update_tx();
```
Comment (janweinstock) [inline src/vcml/models/serial/sifive_uart.cpp:151]: `txctrl` should also handle setting number of stop bits (bit 1 in `txctrl`)

File: src/vcml/models/serial/sifive_uart.cpp [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    ie.on_write(&sifive_uart::write_ie);
+
+    ip.sync_never();
+    ip.allow_read_only();
+
+    div.sync_never();
+    div.allow_read_write();
+    div.on_write(&sifive_uart::write_div);
+
+    serial_tx.set_baud(SERIAL_115200BD);
```
Comment (janweinstock) [inline src/vcml/models/serial/sifive_uart.cpp:219]: serial_tx is missing some setup:
* setting data bits to 8
* setting parity to none

**Review by janweinstock — CHANGES_REQUESTED** [review]

File: src/vcml/models/serial/sifive_uart.cpp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+u32 sifive_uart::read_rxdata() {
+    u32 val = RXDATA_EMPTY;
+    if (is_rx_enabled() && !is_rx_empty()) {
+        val = m_rx_fifo.front();
+        m_rx_fifo.pop();
+        update_rx();
+    }
+    return val;
```
[the remark this answers is not in the document]
Comment (RubBra) [reply src/vcml/models/serial/sifive_uart.cpp]: on further inspection the enabled check doesn't seem to be necessary. disabling rx should only disable enqueuing not dequeuing

**Review by janweinstock — APPROVED** [review]

**Pull request merged**